lib/http-signatures/src/cavage/safety_check.rs

```diff
@@ -0,0 +1,108 @@
+use http::{header::DATE, Method, Request};
+use std::time::{Duration, SystemTime, SystemTimeError};
+use thiserror::Error;
+
+use super::SignatureHeader;
+
+/// 15 minutes
+const MAX_ACCEPTED_SIGNATURE_AGE: Duration = Duration::from_secs(15 * 60);
+
+const REQUIRED_GET_HEADERS: &[&str] = &["host"];
+const REQUIRED_POST_HEADERS: &[&str] = &["host", "content-type", "digest"];
+
+#[derive(Debug, Error)]
+pub enum SafetyCheckError {
+    #[error(transparent)]
+    InvalidDateHeader(#[from] httpdate::Error),
+
+    #[error(transparent)]
+    InvalidHeaderValue(#[from] http::header::ToStrError),
+
+    #[error(transparent)]
+    InvalidSystemTime(#[from] SystemTimeError),
+
+    #[error("Missing required headers")]
+    MissingRequiredHeaders,
+
+    #[error("Signature too old")]
+    SignatureTooOld,
+
+    #[error("Unsupported HTTP method")]
+    UnsupportedHttpMethod,
+}
+
+#[inline]
+fn is_subset<I>(left: &[I], right: &[I]) -> bool
+where
+    I: PartialEq,
+{
+    if left.len() <= right.len() {
+        left.iter().all(|item| right.contains(item))
+    } else {
+        false
+    }
+}
+
+pub fn is_safe<'a, B, I, S>(
+    req: &Request<B>,
+    signature_header: &SignatureHeader<'_, I, S>,
+) -> Result<(), SafetyCheckError>
+where
+    I: Iterator<Item = &'a str> + Clone,
+{
+    let collected_headers = signature_header.headers.clone().collect::<Vec<&str>>();
+    let is_subset = match *req.method() {
+        Method::GET => is_subset(REQUIRED_GET_HEADERS, &collected_headers),
+        Method::POST => is_subset(REQUIRED_POST_HEADERS, &collected_headers),
+        _ => return Err(SafetyCheckError::UnsupportedHttpMethod),
+    };
+
+    if !is_subset {
+        return Err(SafetyCheckError::MissingRequiredHeaders);
+    }
+
+    // Check if the `headers` field either includes `date` or `(created)`
+    if !collected_headers.contains(&"date") && !collected_headers.contains(&"(created)") {
+        return Err(SafetyCheckError::MissingRequiredHeaders);
+    }
+
+    if let Some(created) = signature_header.created {
+        let created_time = SystemTime::UNIX_EPOCH + Duration::from_secs(created);
+        let now = SystemTime::now();
+
+        if now.duration_since(created_time)? > MAX_ACCEPTED_SIGNATURE_AGE {
+            return Err(SafetyCheckError::SignatureTooOld);
+        }
+    }
+
+    if let Some(date_header) = req.headers().get(DATE) {
+        let date_header_time = httpdate::parse_http_date(date_header.to_str()?)?;
+        let now = SystemTime::now();
+
+        if now.duration_since(date_header_time)? > MAX_ACCEPTED_SIGNATURE_AGE {
```

Comment:
Potential edge case: `(created)` + `Date`. Which should we trust? 

Also, before checking it, I think it should be checked signature includes it or not.

Reply:
I think... we can use `min((created), Date)` in here, This should be counted on the calculation of `(expires) - (signature created timestamp)`.

Also, it should be checked against time-traveling scenario. (We should not accept request from the future)

Reply:
Even though.. worst case you could only get through one of the checks without getting your signature rejected.  
Let's say you have a `Date` header which is correct and a `(created)` field that is attacker controlled.

Then the first check, the `(created)` check, might work through the attacker adding an invalid timestamp, but the signature will be discarded when checking the `Date` header (or vice versa).

So I'd say the current code is logically sound, even if one check is technically redundant.

Reply:
Unless one attack vector one wants to proof against is the malicious modification of incoming activities to break federation for you. But I think that attack vector is a bit out there since it requires a MITM position not only in your data centre's packets but also in your TLS connection with other instances.